### PR TITLE
49297 connection resilience

### PR DESCRIPTION
**Pull request by tomblench**
_What_: Ensure pull replication reliability when transferring data over unreliable networks.

_Why_: A quick code audit after HTTP layer replacement shows that our current ability to recover from transient network errors is poor.

_How_: Modify all CouchClient methods to retry when they fail due to a transient error (500-599 error codes, socket closed, connection reset). This is the job of the new `ExecuteResult` class.

All payloads have to be capable of being resent. In the case of `String` or `byte[]` this is easy because the payload is held in memory. For `InputStream`s they must be re-generated with each attempt. This is the purpose of the `InputStreamGenerator` class. `MultipartAttachmentWriter` has been re-written to hold references to the attachments and body and (re-)generate the stream on demand, through `makeInputStream` and `makeInputStreamGenerator`.

_Testing_: Add `UnreliableNetwork{Pull,Push}Test` class and `unreliableNetworkTest` gradle target. Running this will require our custom fork of `toxiproxy` to be running, but the test performs all proxy configuration. The `toxiproxy` fork can be found at https://github.com/tomblench/toxiproxy 

reviewer: @mikerhodes 
reviewer: @rhyshort 
reviewer: @ricellis

### CODE REVIEW

File: sync-core/src/main/java/com/cloudant/mazha/CouchClient.java [hunk 1]
```diff
@@ -180,11 +218,16 @@ public ChangesResult changes(String filterName, Map<String, String> filterParame
...
-    public ChangesResult changes(Map<String, Object> options) {
-        Preconditions.checkNotNull(options, "options must not be null");
-        URI changesFeedUri = this.uriHelper.changesUri(options);
-        HttpConnection connection = Http.GET(changesFeedUri);
-        return executeToJsonObject(connection, ChangesResult.class);
+    public ChangesResult changes(final Map<String, Object> options) {
+        return executeWithRetry(new Callable<ChangesResult>() {
+            @Override
+            public ChangesResult call() throws Exception {
+                Preconditions.checkNotNull(options, "options must not be null");
```
Comment (rhyshort) [inline sync-core/src/main/java/com/cloudant/mazha/CouchClient.java]: I think we should do the `Preconditions` before calling `executeWithRetry` call.

File: sync-core/src/main/java/com/cloudant/mazha/CouchClient.java [hunk 2]
```diff
@@ -551,6 +613,7 @@ private String createBulkSerializedDocsPayload(List<String> serializedDocs) {
                 return diff;
             } catch (IOException ioe) {
                 // return empty map
+                // TODO
```
Comment (rhyshort) [inline sync-core/src/main/java/com/cloudant/mazha/CouchClient.java]: Unneeded comment?

**Comment by rhyshort** [comment]
New Files are missing copyright statements

File: sync-core/src/test/java/com/cloudant/common/ProxyTestBase.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * Created by tomblench on 23/07/15.
+ */
+public class ProxyTestBase extends ReplicationTestBase {
+
+    // test.couch.port becomes proxyPort 'this is where toxiproxy listens'
+    // test.couch.proxy.admin.port
+    // test.couch.proxy.target.port
+
+    // TODO set these up through config
```
Comment (ricellis) [inline sync-core/src/test/java/com/cloudant/common/ProxyTestBase.java]: I think you should add `System.getProperty("test.couch.port",8000)` type calls here as the properties already appear to be in the gradle. Although I also think the properties should be customizable in the gradle following the usual pattern.

File: sync-core/src/main/java/com/cloudant/mazha/CouchClient.java [hunk 4]
```diff
@@ -120,6 +125,34 @@ private InputStream executeToInputStream(HttpConnection connection) throws Couch
...
+            // unpack cause from RetryException
+            // if it's a CouchException, throw as-is
+            // anything else, return null
+            try {
+                throw re.getCause();
+            } catch (CouchException ce) {
+                logger.fine("Re-throwing CouchException from executeWithRetry: "+ce);
+                throw ce;
+            } catch (Throwable e) {
+                return null;
```
Comment (ricellis) [inline sync-core/src/main/java/com/cloudant/mazha/CouchClient.java]: Do we need to log these non-`CouchException` cases? Even if they are unlikely it might be worth having a warning in the log or something so we can figure out why a null might have come back.

File: sync-core/src/main/java/com/cloudant/common/RetriableTask.java [hunk 5]
```diff
@@ -70,6 +70,10 @@ public T call() throws Exception {
                     logger.fine("Success after retry: " + triesRemaining + ", " + task.toString());
```
Comment (ricellis) [inline sync-core/src/main/java/com/cloudant/common/RetriableTask.java]: I realise this is unrelated to your current change so maybe should be another bug, but it is the first time I've seen this code:
`triesRemaining < 3` is wrong it should be `triesRemaining < totalRetries`
Also the logging messages are wrong, they should be
`logger.fine("Retry #: " + (totalRetries - triesRemaining) + "," + task.toString());`
and
`logger.fine("Success after retry: " + (totalRetries - triesRemaining) + ", " + task.toString());`
or maybe be reworded to be about tries remaining to potentially avoid the confusion of "Retry #0" being the first try.

Comment (tomblench) [inline sync-core/src/main/java/com/cloudant/common/RetriableTask.java]: Fixed in 0a97d00

File: sync-core/src/main/java/com/cloudant/mazha/CouchClient.java (hunk 1 above)
Comment (tomblench) [inline sync-core/src/main/java/com/cloudant/mazha/CouchClient.java]: Fixed in 6dfbc8a

File: sync-core/src/test/java/com/cloudant/common/ProxyTestBase.java (hunk 3 above)
Comment (tomblench) [inline sync-core/src/test/java/com/cloudant/common/ProxyTestBase.java]: Done in  013e4317176c33fffa807717f4d0d8641aada924

**Comment by tomblench** [comment]
@rhyshort copyright headers added in 540b568

File: sync-core/src/main/java/com/cloudant/common/RetriableTask.java [hunk 6]
```diff
@@ -62,26 +60,27 @@ public int getTriesRemaining() {
...
-                            "ms interval", e);
-                }
-                logger.log(Level.FINE,"Retry later: " + triesRemaining + ", " + task.toString(), e);
+                lastException = e;
                 Thread.sleep(timeToWait);
             }
-        }
+        } while (--triesRemaining > 0);
+        // if we got here, we ran out of retries
+        throw new RetryException(String.format("Failed after %d attempts to remaining for task %s",
```
Comment (ricellis) [inline sync-core/src/main/java/com/cloudant/common/RetriableTask.java]: The exception message doesn't make sense, should it just say "`Failed after %d attempts for task %s"`

**Comment by ricellis** [comment]
+1 once the message in https://github.com/cloudant/sync-android/pull/189#discussion-diff-36870549 is fixed

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/common/RetriableTask.java [hunk 7]
```diff
@@ -62,26 +60,27 @@ public int getTriesRemaining() {
...
+        do {
             try {
-                logger.fine("Retry #: " + triesRemaining + "," + task.toString());
+                logger.fine(String.format("%d tries remaining for task %s", triesRemaining, task));
                 T t = task.call();
-                if(triesRemaining < 3) {
-                    logger.fine("Success after retry: " + triesRemaining + ", " + task.toString());
-                }
+                logger.fine(String.format("Success with %d tries remaining for task %s",
+                        triesRemaining, task));
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/common/RetriableTask.java]: I think there is an off by one issue here because the successful attempt has consumed one of the tries. In which case the log message should be `triesRemaining - 1`. Although arguably since the field and parameter are called `totalRetries` not `totalTries` then the message is actually correct and the `while` condition should be `--triesRemaining >= 0`.

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/common/RetriableTask.java [hunk 8]
```diff
@@ -62,26 +60,27 @@ public int getTriesRemaining() {
...
-                }
-                logger.log(Level.FINE,"Retry later: " + triesRemaining + ", " + task.toString(), e);
+                lastException = e;
                 Thread.sleep(timeToWait);
             }
-        }
+        } while (--triesRemaining > 0);
+        // if we got here, we ran out of retries
+        throw new RetryException(String.format("Failed after %d attempts for task %s",
+                totalRetries, task), lastException);
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/common/RetriableTask.java]: Depending on how the previous comment is addressed this might need to be `totalRetries + 1` attempts.

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/http/HttpConnection.java [hunk 9]
```diff
@@ -117,7 +117,12 @@ public HttpConnection setNumberOfRetries(int numberOfRetries){
      * @return an {@link HttpConnection} for method chaining 
      */
     public HttpConnection setRequestBody(final String input) {
-        this.input = new ByteArrayInputStream(input.getBytes());
+        this.input = new InputStreamGenerator() {
+            @Override
+            public InputStream getInputStream() {
+                return new ByteArrayInputStream(input.getBytes());
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/http/HttpConnection.java]: This uses the platforms default charset, I think we should probably use the more explicit `getBytes("UTF-8")`.

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/http/HttpConnection.java [hunk 10]
```diff
@@ -117,7 +117,12 @@ public HttpConnection setNumberOfRetries(int numberOfRetries){
...
+            }
+        };
         // input is in bytes, not characters
         this.inputLength = input.getBytes().length;
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/http/HttpConnection.java]: The charset might have an impact on this size, so so should probably be specified again.

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java [hunk 11]
```diff
@@ -68,22 +72,92 @@ public URI getRootUri() {
         return this.uriHelper.getRootUri();
     }
 
+    // result of executing an HTTP call:
+    // - stream non-null and exception null: the call was successful, result in stream
+    // - stream null and exception non-null: the call was unsuccessful, details in exception
+    // - fatal: set to true when exception non-null, indicates call should not be retried
+    private class ExecuteResult
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java]: Perhaps there is an argument for this to be extracted to a top level class and potentially used as the return type for the `HttpConnection#execute()`. Although I consider that refactor out of scope for this PR.

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java [hunk 12]
```diff
@@ -68,22 +72,92 @@ public URI getRootUri() {
...
+                    // 1xx and 2xx are OK
+                    // we would not normally expect to see 1xx in a response
+                    // explicitly set these to show we are OK
+                    this.fatal = false;
+                    this.exception = null;
+                    this.stream = stream;
+                    break;
+                case 3:
+                    // 3xx redirection
+                    // TODO?
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java]: is there anything TODO?

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java [hunk 13]
```diff
@@ -68,22 +72,92 @@ public URI getRootUri() {
...
+        private ExecuteResult(InputStream stream,
+                              InputStream errorStream,
+                              int responseCode,
+                              String responseMessage,
+                              Throwable cause,
+                              JSONHelper jsonHelper)
+        {
+            this.responseCode = responseCode;
+            boolean needsCouchException = false;
+            System.out.println("Response: "+responseMessage);
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java]: Log or delete

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java [hunk 14]
```diff
@@ -68,22 +72,92 @@ public URI getRootUri() {
...
+    }
+
     // - if 2xx then return stream
     // - map 404 to NoResourceException
     // - if there's a couch error returned as json, un-marshall and throw
     // - anything else, just throw the IOException back, use the cause part of the exception?
 
     // it needs to catch eg FileNotFoundException and rethrow to emulate the previous exception handling behaviour
-    private InputStream executeToInputStream(HttpConnection connection) throws CouchException {
+    private ExecuteResult executeToInputStream(HttpConnection connection) {
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java]: Method name is a bit strange now the return type is not an `InputStream`

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java [hunk 15]
```diff
@@ -91,32 +165,35 @@ private InputStream executeToInputStream(HttpConnection connection) throws Couch
...
+            ExecuteResult executeResult = new ExecuteResult(inputStream,
+                    errorStream,
+                    responseCode,
+                    responseMessage,
+                    cause,
+                    jsonHelper);
+            return executeResult;
         } finally {
-            if (es != null) {
+            if (errorStream != null) {
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java]: Unfair to call this up when all you did was rename the field to something that made more sense, but wouldn't this be a good place for `IOUtils.close()`?

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java [hunk 16]
```diff
@@ -124,16 +201,65 @@ private InputStream executeToInputStream(HttpConnection connection) throws Couch
         }
     }
 
-    private <T> T executeToJsonObject(HttpConnection connection, Class<T> c) throws CouchException {
-        InputStream is = this.executeToInputStream(connection);
+    // execute HTTP task with retries:
+    // return an InputStream if successful or throw an exception
+    private InputStream executeWithRetry(final Callable<ExecuteResult> task) throws CouchException {
+        int retries = 10;
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java]: Again I find the `retries` terminology slightly misleading. As coded these are `tries` or `attempts`, but `retries` suggests the while loop should be `>= 0` again.

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java [hunk 17]
```diff
@@ -124,16 +201,65 @@ private InputStream executeToInputStream(HttpConnection connection) throws Couch
...
+                                               Class<T> c) throws CouchException {
+        InputStream is = this.executeToInputStreamWithRetry(connection);
         InputStreamReader isr = new InputStreamReader(is);
-        T json = new JSONHelper().fromJson(isr, c);
-        return json;
+        try {
+            T json = new JSONHelper().fromJson(isr, c);
+            return json;
+        } finally {
+            try {
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java]: `IOUtils.close()`?

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java [hunk 18]
```diff
@@ -124,16 +201,65 @@ private InputStream executeToInputStream(HttpConnection connection) throws Couch
...
+        });
+        return is;
     }
 
     public void createDb() {
-        HttpConnection connection = Http.PUT(this.uriHelper.getRootUri(), "application/json");
-        DBOperationResponse res = executeToJsonObject(connection, DBOperationResponse.class);
+        HttpConnection connection = Http.PUT(uriHelper.getRootUri(), "application/json");
+        DBOperationResponse res = executeToJsonObjectWithRetry(connection, DBOperationResponse.class);
+        // TODO this should now be unreachable
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java]: What needs doing? Delete the next block?

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java [hunk 19]
```diff
@@ -141,18 +267,20 @@ public void createDb() {
 
     public void deleteDb() {
         HttpConnection connection = Http.DELETE(this.uriHelper.getRootUri());
-        DBOperationResponse res = executeToJsonObject(connection, DBOperationResponse.class);
+        DBOperationResponse res = executeToJsonObjectWithRetry(connection, DBOperationResponse
+                .class);
+        // TODO this should now be unreachable
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java]: Again what needs doing here?

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/datastore/MultipartAttachmentWriter.java [hunk 20]
```diff
@@ -154,81 +159,15 @@ public void addAttachment(Attachment attachment, long length) throws IOException
      * This <b>must</b> be called before any client attempts to issue a read.
      */
     public void close() {
+        // TODO we could just remove this because we always know the byte count will include this
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/datastore/MultipartAttachmentWriter.java]: Is the `trailingBoundary` length encoding dependent?

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/datastore/MultipartAttachmentWriter.java [hunk 21]
```diff
@@ -272,6 +211,113 @@ public String getRevision() {
...
+        for (Attachment a : attachments) {
+            try {
+                components.add(new ByteArrayInputStream(crlf));
+                components.add(new ByteArrayInputStream(partBoundary));
+                components.add(new ByteArrayInputStream(crlf));
+                components.add(new ByteArrayInputStream(crlf));
+                components.add(a.getInputStream());
+            } catch (IOException ioe) {
+                // TODO - this should set a flag which defers throwing ioexception when read() is called on the InputStream
+                throw new RuntimeException("aargh", ioe);
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/datastore/MultipartAttachmentWriter.java]: Probably need a better message here :)

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/common/RetriableTask.java (hunk 8 above)
Comment (tomblench) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/common/RetriableTask.java]: @ricellis see af832a44f9977f313791206d9fe4803fbd7f9cae for this and the above comment. I've changed retries to tries in the appropriate places because that's what the tests imply. 

I also removed the "Success" message since I don't think it adds anything useful.

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java (hunk 11 above)
Comment (tomblench) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java]: Agreed this could do with more refactoring in another PR.

**Comment by ricellis** [comment]
+1

**Comment by rhyshort** [comment]
Is there some doc about which version of Toxiproxy is needed? It looked  like it needs version 2 thats in dev (last time I checked)

**Comment by tomblench** [comment]
@rhyshort from the PR description:

_Running this will require our custom fork of toxiproxy to be running, but the test performs all proxy configuration. The toxiproxy fork can be found at https://github.com/tomblench/toxiproxy_

The toxiproxy project accepted the suggestions in my PR - https://github.com/Shopify/toxiproxy/pull/65 - but they haven't released 2.0 yet.

**Comment by rhyshort** [comment]
Ah shows how long its been since I read the description, +1
